zerver/data_import/mattermost.py

```diff
@@ -0,0 +1,540 @@
+
+    realm_id = 0
+    for team in mattermost_data["team"]:
+        team_name = team["name"]
+
+        user_handler = UserHandler()
+        subscriber_handler = SubscriberHandler()
+        user_id_mapper = IdMapper()
+        stream_id_mapper = IdMapper()
+
+        print("Generating data for", team_name)
+        realm = make_realm(realm_id, team)
+        realm_output_dir = os.path.join(output_dir, team_name)
+
+        reset_mirror_dummy_users(username_to_user)
+        label_mirror_dummy_users(team_name, mattermost_data, username_to_user)
+
+        convert_user_data(
+            user_handler=user_handler,
+            user_id_mapper=user_id_mapper,
+            user_data_map=username_to_user,
+            realm_id=realm_id,
+            team_name=team_name,
+        )
+
+        zerver_stream = convert_channel_data(
+            channel_data=mattermost_data["channel"],
+            user_data_map=username_to_user,
+            subscriber_handler=subscriber_handler,
+            stream_id_mapper=stream_id_mapper,
+            user_id_mapper=user_id_mapper,
+            realm_id=realm_id,
+            team_name=team_name,
+        )
+
+        realm['zerver_stream'] = zerver_stream
+
+        all_users = user_handler.get_all_users()
+
+        zerver_recipient = build_recipients(
+            zerver_userprofile=all_users,
+            zerver_stream=zerver_stream,
+        )
+        realm['zerver_recipient'] = zerver_recipient
+
+        stream_subscriptions = build_stream_subscriptions(
+            get_users=subscriber_handler.get_users,
+            zerver_recipient=zerver_recipient,
+            zerver_stream=zerver_stream,
+        )
+
+        personal_subscriptions = build_personal_subscriptions(
+            zerver_recipient=zerver_recipient,
+        )
+
+        zerver_subscription = personal_subscriptions + stream_subscriptions
+        realm['zerver_subscription'] = zerver_subscription
+
+        subscriber_map = make_subscriber_map(
+            zerver_subscription=zerver_subscription,
+        )
+
+        write_message_data(
+            team_name=team_name,
+            realm_id=realm_id,
+            post_data=mattermost_data["post"],
+            zerver_recipient=zerver_recipient,
+            subscriber_map=subscriber_map,
+            output_dir=realm_output_dir,
+            masking_content=masking_content,
+            stream_id_mapper=stream_id_mapper,
+            user_id_mapper=user_id_mapper,
+            user_handler=user_handler,
+            username_to_user=username_to_user,
+        )
+
+        realm['zerver_userprofile'] = user_handler.get_all_users()
+        realm['sort_by_date'] = True
+
+        create_converted_data_files(realm, realm_output_dir, '/realm.json')
+        create_converted_data_files([], realm_output_dir, '/avatars/records.json')
+        create_converted_data_files([], realm_output_dir, '/uploads/records.json')
```

Comment:
Let's generate an empty `/emoji/records.json` file here, and add a comment for each of the empty records we're creating in this block indicating which ones Mattermost supports (but aren't supported yet).  

Though I think it's probably worth taking the time to setup a quick Mattermost server, upload a custom emoji, and just make that work, since it's probably pretty quick and then we don't need to document the missing feature.

Reply:
I created a Mattermost server using docker and created export data dump with custom emojis. Will make the tool export emoji as well.